intel_adsp: Introduce debug slot manager

ujfalusi opened the pull request:
Currently the debug slots are 'allocated' in wild-west style, no real rules enforced, one has to know exactly which module uses what slots (in hard-wired fashion mostly) which can lead to races and bugs that slots have multiple users and at the same time we still have empty, unused slot(s).

The debug slot manager provides a way for drivers to request a debug slot for their use, or (in case of panic handler) forcibly overtake a given slot.

The debug slot manager can be enabled with a Kconfig change to avoid breakage in compilation - SOF also have modules which tampers with the descriptor page and slots directly.

ujfalusi committed:
intel_adsp: debug_window: Describe the partial slot in page0

Resize the reserved part of the debug window to cover the first 1K and
define the partial_page0 which will cover the partial slot in page 0.

Add comment to describe the debug window layout.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
soc: intel_adsp: tools: cavstool.py: Look up the shell slot by type

Instead of using hardwired index for the shell debug slot, look it up by
it's type.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
intel_adsp: kconfig: Increase the default size of MEMORY_WIN_2_SIZE

The current default of 8192 will provide only 2 pages in debug window:
page0: descriptors
page1: slot0

However, the coredump is hardwired to use slot1, which by default is not
valid.

Increase the default window size to 12288 to allow three pages.

This change affects CAVS25 only as it is using default window sizes and the
window 3 is not used in this configuration at all (it was used with IPC3
only), so we do have enough space for the three page - we could even
increase the default to cover 4 pages (8192+8192), but let's be
conservative on this.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
intel_adsp: Add debug slot manager

Currently all drivers which uses a slot from the debug window have fragile
hardwired slot 'mapping', they are locked to use specific slots even if
there are free slots available for them to take.

The new API hides the management of the slots and descriptors and users
can ask, release or even seize slots that they want to use.

Add a new debug slot manager API and a new default no config option to
allow selection between the hardwired or dynamic debug slot management.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

abonislawski commented on soc/intel/intel_adsp/common/debug_window.c:
```diff
@@ -0,0 +1,158 @@
+		}
+	}
+
+	ADSP_DW->descs[slot_idx].resource_id = dw_desc->resource_id;
```
So if adsp_dw_find_slot returns a valid slot_idx, this function will still overtake this slot and overwrite resource_id and vma even if the previous user did not release it? Is this intentional? Because looks like a task intended for adsp_dw_seize_slot.

abonislawski reviewed (commented):
In high level looks like a nice upgrade. I verified mtrace and at least it still works :)
Just a question about request_slot, it looks suspicious:

nashif commented on soc/intel/intel_adsp/common/debug_window.c:
```diff
@@ -0,0 +1,176 @@
+	slot_idx = adsp_dw_find_slot_by_type(dw_desc->type);
+	if (slot_idx >= 0) {
+		return ADSP_DW->slots[slot_idx];
+	} else {
```
are you going to address this?

ujfalusi commented:
Changes since v6:
- remove unnecessary else from adsp_dw_request_slot()

kartben commented on subsys/debug/coredump/coredump_backend_intel_adsp_mem_window.c:
```diff
@@ -46,10 +61,20 @@ static void coredump_mem_window_backend_end(void)
 	size_t data_left;
+#ifdef CONFIG_INTEL_ADSP_DEBUG_SLOT_MANAGER
+	uint32_t *mem_window_separator = (uint32_t *)coredump_slot_addr;
+	uint8_t *mem_window_sink = (uint8_t *)coredump_slot_addr + 4 + mem_wptr;
```
+1, rule 11.8

kartben reviewed (changes requested):
please address the coding guidelines issues as per earlier comment - thanks!

ujfalusi committed:
intel_adsp: common: gdbstub: Correct z_gdb_backend_init() return type

The z_gdb_backend_init() is defined to return with success/error code
with return type of int.

Fixes: a9c47a47a4f3 ("intel_adsp: cavs: add gdb support")
Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v8:
- shorten the commit title of the first patch
